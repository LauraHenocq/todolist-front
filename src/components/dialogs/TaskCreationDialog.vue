<script setup lang="ts">
import TaskForm from '@/components/forms/TaskForm.vue';
import type { CreateTaskInput } from "@/services/inputs/create-task.input";
import { computed, ref, watch } from 'vue';

type Props = {
  isOpen: boolean
}

const { isOpen } = defineProps<Props>()

const formRef = ref(null);
const isDialogOpen = ref(isOpen);

watch(() => isOpen, (newVal) => {
  isDialogOpen.value = newVal;
});

const submitForm = () => {
  if (formRef.value) {
    formRef.value?.submitForm();
  }
};

const emit = defineEmits(['create-task', 'cancel']);

const cancel = () => {
  emit('cancel');
}

const createTask = (newTask: CreateTaskInput) => {
  emit('create-task', newTask);
  isDialogOpen.value = false;
}

</script>

<template>
  <v-dialog data-cy="task-creation-dialog" v-model="isDialogOpen" width="90%">
      <v-card title="Créer une nouvelle tâche">
        <TaskForm is-edition-mode="false" ref="formRef" @createTask="createTask"/>

        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn
            data-cy="cancel-creation-task-button"
            text="Annuler"
            @click="cancel"
          ></v-btn>
          <v-btn
            data-cy="validate-creation-task-button"
            text="Valider"
            @click="submitForm"
          ></v-btn>
        </v-card-actions>
      </v-card>
  </v-dialog>
</template>