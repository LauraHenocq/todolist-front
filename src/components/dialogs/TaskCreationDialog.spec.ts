import { mount, VueWrapper } from '@vue/test-utils';
import { nextTick } from 'vue';
import { beforeEach, expect, it, describe, vi } from 'vitest'
import { createVuetify } from 'vuetify';
import * as components from 'vuetify/components'
import * as directives from 'vuetify/directives'
import TaskCreationDialog from './TaskCreationDialog.vue';
import 'vuetify/styles'; 

const vuetify = createVuetify({
  components,
  directives
});

describe('TaskCreationDialog', () => {
  let wrapper: VueWrapper
  const isOpen = true;

  const defaultProps = {
    isOpen
  }

  global.ResizeObserver = require('resize-observer-polyfill')

  beforeEach(() => {
    wrapper = mount(TaskCreationDialog, {
      props: defaultProps,
      global: {
        plugins: [vuetify],
      },
    });
  });

  describe('render', () => {
    it('should render', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe.skip('actions', () => {
    it('should emit create-task event when the createTask event is emitted', async () => {
      (wrapper
        .findComponent('[data-cy="task-card-edit-task-button"]') as VueWrapper)
        .vm.$emit('createTask');

      await nextTick();

      expect(wrapper.emitted('create-task')).toHaveLength(1);
    });

    it('should submit the form when the validate task button is clicked', async () => {
      const submitFormSpy = vi.spyOn(wrapper.vm, 'submitForm');
      (wrapper
        .findComponent('[data-cy="validate-creation-task-button"]') as VueWrapper)
        .vm.$emit('click');

        await nextTick();

      expect(submitFormSpy).toHaveBeenCalled();
    });

    it('should emit cancel event when the cancel button is clicked', async () => {
        (wrapper
          .findComponent('[data-cy="cancel-creation-task-button"]') as VueWrapper)
          .vm.$emit('click');
  
          await nextTick();
  
        expect(wrapper.emitted('delete-task')).toHaveLength(1);
      });
  });
});