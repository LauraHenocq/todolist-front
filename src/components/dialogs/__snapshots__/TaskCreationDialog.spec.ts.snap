// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskCreationDialog > render > should render 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;
